### PR TITLE
HSM Docs

### PR DESCRIPTION
**Pull request by nklaassen**

### CODE REVIEW

File: docs/pages/enterprise/hsm.mdx [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+description: How to configure Hardware Security Modules to manage your Teleport CA private keys
+h1: Teleport HSM Support
+---
+
+Teleport can be configured to use a connected HSM to create and hold CA private
+keys.
+
+While most PKCS#11 HSMs should be supported, we test with AWS CloudHSM,
+YubiHSM2, and SoftHSM2.
+
```
Comment (klizhentas) [inline docs/pages/enterprise/hsm.mdx:12]: Please follow the structure similar to the articles here:

https://github.com/gravitational/teleport/blob/master/docs/pages/setup/guides/fluentd.mdx

## Prerequisites

- Teleport v(=teleport.version=) Cloud, Open Source or Enterprise
- Teleport admin tool `tctl` configured to access the cluster.
- Fluentd version v(=fluentd.version=).
- Docker version v(=docker.version=).

```code
$ terraform version
# Terraform v(=terraform.version=)
```

```code
$ tctl version
# Teleport v(=teleport.version=) go(teleport.golang)
#
# Test connectivity and admin permissions
$ tctl get roles
```

Create a folder `fluentd` to hold configuration and plugin state:

```code
$ mkdir -p event-handler
$ cd event-handler
```

## Step 1/n. Do this

File: docs/pages/enterprise/hsm.mdx [hunk 2]
```diff
@@ -0,0 +1,205 @@
+---
```
Comment (klizhentas) [inline docs/pages/enterprise/hsm.mdx:1]: Please consult the meta docs on docs to follow the style guide:

https://goteleport.com/docs/docs/

File: docs/pages/enterprise/hsm.mdx [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+title: Teleport HSM Support
+description: How to configure Hardware Security Modules to manage your Teleport CA private keys
+h1: Teleport HSM Support
+---
+
+Teleport can be configured to use a connected HSM to create and hold CA private
+keys.
+
+While most PKCS#11 HSMs should be supported, we test with AWS CloudHSM,
```
Comment (klizhentas) [inline docs/pages/enterprise/hsm.mdx:10]: The teleport team tests...

File: docs/pages/enterprise/hsm.mdx [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+YubiHSM2, and SoftHSM2.
+
+# Setup
```
Comment (klizhentas) [inline docs/pages/enterprise/hsm.mdx:13]: Remove `#` and Setup in favor of prerequisites. Also we should not have `#` only `##`, as `#` is H1.

File: docs/pages/enterprise/hsm.mdx [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+This guide assumes that you are familiar with setting up Teleport. HSMs need to
+be configured on all of your Auth servers. Each Auth server can be connected to
+a unique HSM, or they can all connect to the same HSM depending on your usecase.
+
+## HSM Setup
```
Comment (klizhentas) [inline docs/pages/enterprise/hsm.mdx:19]: To make sure the article flows from top to bottom add examples inline and use TabItem elements to let users pick the type of the HSM setup.

File: docs/pages/enterprise/hsm.mdx [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+You will need to set up your HSM and make sure that it is accessible from your
+Teleport Auth server. You should create a unique HSM user or token for Teleport
+to use. Install your HSM vendor's PKCS#11 library on the machine where Teleport
+will be running. See [Examples](#Examples) for basic instructions on setting up
+some common HSMs.
+
+## Teleport Configuration
```
Comment (klizhentas) [inline docs/pages/enterprise/hsm.mdx:27]: Should be `## Step 3/n. Configure Auth Server`

File: docs/pages/enterprise/hsm.mdx [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+      # depending on your HSM and library, slot_number (CK_SLOT_ID) can
+      # optionally be used to select the token instead of token_label
+      # slot_number: 1
+
+      # pin should be the token pin, the format will depend on your HSM and
+      # PKCS#11 library.
+      pin: "username:password"
+      # pin_path can optionally be used to read the pin from a file
+      # pin_path: /path/to/pin_file
+```
```
Comment (klizhentas) [inline docs/pages/enterprise/hsm.mdx:58]: Should we restart?

File: docs/pages/enterprise/hsm.mdx [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
```
Comment (klizhentas) [inline docs/pages/enterprise/hsm.mdx:59]: How to test that it's working? Troubleshooting?

File: docs/pages/enterprise/hsm.mdx [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+# Certificate Rotation
```
Comment (klizhentas) [inline docs/pages/enterprise/hsm.mdx:60]: That should be in Next steps section after the final step.

```
## Next steps

### Certificate authority rotation with HSM

```

File: docs/pages/enterprise/hsm.mdx [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+HSM-connected Auth server to an HA Teleport cluster, CA rotation needs to be
+performed in order for the cluster to trust the new HSM certificates. `teleport`
+will print a warning during startup if this needs to be completed, and will not
+sign any certificates (except the `Admin` certificate used by `tctl` which will
+be signed by a temporary HSM key). A warning will also be printed in `tctl
+status` if this is required for any Auth server in the cluster.
+
+CA rotation can be performed manually or automatically, see our admin guide on
+[Certificate
+rotation](https://goteleport.com/docs/admin-guide/#certificate-rotation). To rotate the CAs
```
Comment (klizhentas) [inline docs/pages/enterprise/hsm.mdx:72]: This should be internal link: `(../setup/admin/ca-rotation.mdx`)

File: docs/pages/enterprise/hsm.mdx [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+connection to be able to receive new certs or they may lose access to the
+cluster and have to re-join.
+
+You may see some warnings or errors in the logs during rotation while the
+servers are reloading, this is expected while the old instance is being
+terminated. The logs should stabilize within 1-2 minutes.
+
+# Examples
+
+## YubiHSM2
```
Comment (klizhentas) [inline docs/pages/enterprise/hsm.mdx:98]: Let's make these tab items and move above

File: docs/pages/enterprise/hsm.mdx (hunk 8 above)
Comment (klizhentas) [reply docs/pages/enterprise/hsm.mdx:59]: The last step should be verifying that it works and then optional next steps

**Review by klizhentas — CHANGES_REQUESTED** [review]

File: docs/pages/enterprise/hsm.mdx [hunk 12]
```diff
@@ -1,36 +1,109 @@
...
+- Setting up a Teleport auth server to use an HSM to store and handle private
+  keys.
 
-# Setup
+## Prerequisites
 
-This guide assumes that you are familiar with setting up Teleport. HSMs need to
-be configured on all of your Auth servers. Each Auth server can be connected to
-a unique HSM, or they can all connect to the same HSM depending on your usecase.
+- Teleport v(=teleport.version=) Open Source or Enterprise (self-hosted).
```
Comment (klizhentas) [inline docs/pages/enterprise/hsm.mdx:14]: @nklaassen this should be Enterprise-only feature. Please confirm and if so, update the docs as well.

File: docs/pages/enterprise/hsm.mdx [hunk 13]
```diff
@@ -40,166 +113,86 @@ auth_service:
...
-sudo /opt/cloudhsm/bin/configure --cmu <IP address>
-/opt/cloudhsm/bin/cloudhsm_mgmt_util /opt/cloudhsm/etc/cloudhsm_mgmt_util.cfg
-> listUsers
-> loginHSM PRECO admin password
-> changePswd PRECO admin <NewPassword>
-> listUsers
-# user `admin` should change from PRECO to CO(Crypto Officer)
-```
+## Next Steps
+### Certificate Rotation with HSM
```
Comment (klizhentas) [inline docs/pages/enterprise/hsm.mdx:166]: @nklaassen, sorry, I misunderstood. After reading this guide again, it seems like CA rotation is a necessary step to kick off HSM. Let's make it Step 4/4 then. The section `Next steps` is usually for optional steps, and this one seems like mandatory.

File: docs/pages/enterprise/hsm.mdx (hunk 12 above)
Comment (nklaassen) [reply docs/pages/enterprise/hsm.mdx]: Right, it should be Enterprise self-hosted only, I've updated the doc. To be clear, currently the license restriction is not in place, it is easy to do on the teleport side but I did not get a chance to update Houston to create the licenses

File: docs/pages/enterprise/hsm.mdx (hunk 13 above)
Comment (nklaassen) [reply docs/pages/enterprise/hsm.mdx:166]: Right, unless you are starting a brand new auth server and cluster it is mandatory, I will make this step 4

File: docs/pages/enterprise/hsm.mdx [hunk 14]
```diff
@@ -40,166 +113,86 @@ auth_service:
...
-    pkcs11:
-      module_path: /opt/cloudhsm/lib/libcloudhsm_pkcs11.so
-      token_label: cavium
-      pin: "<CU_username>:<CU_password>"
-      # pin_path can optionally be used to read the pin from a file
-      # pin_path: /path/to/pin_file
-```
+If you are updating a live cluster, make sure to wait for all nodes which may
+intermittently lose connection to connect and receive new certs or they may lose
+access to the cluster and have to re-join.
```
[the remark this answers is not in the document]
Comment (nklaassen) [reply docs/pages/enterprise/hsm.mdx:197]: Added a brief `Step 5. Confirm that Teleport is using your HSM`

File: docs/pages/enterprise/hsm.mdx (hunk 12 above)
Comment (nklaassen) [reply docs/pages/enterprise/hsm.mdx]: PRs are up to enforce the license requirement

**Comment by nklaassen** [comment]
Thanks for the guidance @klizhentas, I had a very rough draft here originally but I'm happier with it now and I think I've addressed your comments

**Review by klizhentas — APPROVED** [review]

**Comment by klizhentas** [comment]
@nklaassen you are good to go, please backport to branch/v7 as well once you get a chance

File: docs/pages/enterprise/hsm.mdx [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+status` if this is required for any Auth server in the cluster.
+
+CA rotation can be performed manually or semi-automatically, see our admin guide
+on [Certificate rotation](../setup/operations/ca-rotation.mdx). To rotate the
+CAs manually you can run:
+
+```code
+$ tctl status
+WARNING: One or more auth servers has a newly added or removed HSM...
+$ tctl auth rotate --manual --phase init
```
Comment (tcsc) [inline docs/pages/enterprise/hsm.mdx:184]: I'm finding it hard to parse this block - maybe some vertical whitespace might help here.
